ghcide/src/Development/IDE/Core/Compile.hs

```diff
@@ -435,6 +436,42 @@ tcRnModule hsc_env tc_helpers pmod = do
 -- anywhere. So we zero it out.
 -- The field is not serialized or deserialised from disk, so we don't need to remove it
 -- while reading an iface from disk, only if we just generated an iface in memory
+--
+
+
+
+-- | See https://github.com/haskell/haskell-language-server/issues/3450
+-- GHC's recompilation avoidance in the presense of TH is less precise than
+-- HLS. To avoid GHC from pessimising HLS, we filter out certain dependency information
+-- that we track ourselves. See also Note [Recompilation avoidance in the presence of TH]
+filterUsages :: [Usage] -> [Usage]
+#if MIN_VERSION_ghc(9,3,0)
+filterUsages = filter $ \case UsageHomeModuleInterface{} -> False
+                              _ -> True
+#else
+filterUsages = id
+#endif
+
+-- | Mitigation for https://gitlab.haskell.org/ghc/ghc/-/issues/22744
+shareUsages :: ModIface -> ModIface
+shareUsages iface = iface {mi_usages = usages}
+  where usages = map go (mi_usages iface)
+        go usg = usg {usg_file_path = fp}
+          where !fp = shareFilePath (usg_file_path usg)
+
+filePathMap :: IORef (HashMap.HashMap FilePath FilePath)
```

Comment:
Correct me if I am wrong, but I thought we want to prefer `Var` over `IORef` because it has better guarantees? E.g., it was easier for us to avoid space leaks?

Reply:
I don't think it really matters here, except that `Var` generally has slightly more overhead.